compiler/src/iree/compiler/Codegen/Common/TileSizeSelection.h

```diff
@@ -15,6 +15,75 @@ namespace mlir::iree_compiler {
 using SizesAndScalableFlags =
     std::pair<SmallVector<int64_t>, SmallVector<bool>>;
 
+using SizeAndScalableFlag = std::tuple<int64_t &, bool &>;
+
+/// A tuple that encapsulates two quantities describing tile sizes:
+///   * regular tile sizes (integers) - that's always required
+///   * scalable tile flags (bool - only used/required for scalable
+///     vectorisation.
+/// Use this wrapper to make sure that both quantities are upated when
+/// manipulating tile sizes.
+struct SizesAndScalableFlagsTuple {
+  SmallVector<int64_t> sizes;
+  SmallVector<bool> flags;
+
+  // Represents a pair of references to a size and a scalable flag at the given
+  // index. Due to various implementation details of vector of bools, it's much
+  // easier to store a reference to a whole container and an index. While this
+  // increases the size of this wrapper, it also simplifies the implementation.
+  struct ReferencePair {
+    SmallVector<int64_t> &sizesVec;
+    SmallVector<bool> &flagsVec;
+    // Index of this pair within the vectors
+    size_t index;
+
+    explicit ReferencePair(const ReferencePair &a) = default;
+    ReferencePair(SmallVector<int64_t> &sizesVecRef,
+                  SmallVector<bool> &boolVectorRef, size_t indexRef)
+        : sizesVec(sizesVecRef), flagsVec(boolVectorRef), index(indexRef) {}
+
+    // Update this pair based on the input integer + bool
+    ReferencePair &operator=(const std::pair<int64_t, bool> &values) {
+      sizesVec[index] = values.first;
+      flagsVec[index] = values.second;
+      return *this;
+    }
+
+    // Update this pair based on the input integer. Assume that the scalable
+    // size is false. This is safe to use in cases where no scalable
+    // vectorisation/tiling is used/supported.
+    ReferencePair &operator=(int64_t size) {
+      sizesVec[index] = size;
+      flagsVec[index] = false;
+      return *this;
+    }
+
+    // Update this pair based on the input ReferencePair
+    ReferencePair &operator=(const ReferencePair &pair) {
+      sizesVec[index] = pair.sizesVec[index];
+      flagsVec[index] = pair.flagsVec[index];
+      return *this;
+    }
+  };
+
+  SizesAndScalableFlagsTuple(SmallVector<int64_t> s, SmallVector<bool> f)
+      : sizes(s), flags(f) {}
+
+  // Initialise to {0, false} for all sizes
+  SizesAndScalableFlagsTuple(size_t numElements)
+      : sizes(SmallVector<int64_t>(numElements, 0)),
+        flags(SmallVector<bool>(numElements, false)) {}
+
+  SizesAndScalableFlags get() {
+    return std::pair<SmallVector<int64_t>, SmallVector<bool>>(sizes, flags);
+  }
+
+  ReferencePair operator[](size_t index) {
```

Comment:
How do we make sure the parent `SizesAndScalableFlagsTuple` will outlive the returned `ReferencePair`?

This allows users to get and store a reference, which seems easily create a bug that keeps using the reference pair while the `SizesAndScalableFlagsTuple` has been destructed

Reply:
Thanks for pointing this out!

Not sure there’s a way around it, TBH. Well, other than replacing (“pair of vectors”)
```cpp
SmallVector<int64_t> commonVecTileSizes = parallelVecTileSizes;
SmallVector<bool> commonVecScalableTileFlags = parallelVecScalableTileSizes;
```

with (“vector of pairs”):
```cpp
SmallVector<ScalableTileSize> commonVecTileSizesAndFlags 
    = parallelVecTileSizesAndFlags
```

That would be more intrusive, but would allow safer code. Having said that, `ReferencePair` is not intended for general use anyway. WDYT?

Reply:
Yes, before I found this change I was thinking about replacing it with vector of pair `SmallVector<ScalableTileSize>` everywhere. And in the end convert it to two separate vectors `tileSizes` and `scalableTileFlags` for the `lowering_config` attribute here:

https://github.com/openxla/iree/blob/9eff8615c2b2615db9f2ba6c79211452329aa872/compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenAttrs.h#L135-L152

Ideally we can also have [{tileSize, scalableFlag}, ...] on `lowering_config` attribute, but that will be a larger refactor and can be done separately

Reply:
+1

I am really glad that you also see this as a better design. 

One small nit, I would call it `TileSizeAndScalableFlag` rather than `ScalableTileSize`. The former indicates that 2 values are encapsulated - that would be always true. The latter suggests that tile sizes are always scalable - not always true - tile sizes can be both fixed width _or_ scalable. Naming is hard.

Reply:
I think we can have a C++ class which maintain all the information about tile sizes and scalable flags. Then we won't be bothered by pair and vector things. The class generates `SmallVector<SmallVector<ScalableTileSize>>` at the end, and we use it to set the strategy and configuration. It changes the whole flow, and it requires a broader review from folks. I proposed the idea internally (in AMD) and collected some feedback from @pzread and @dcaballe, but I did not get a chance to push it forward. I'll try to find some cycles and can share with you offline.

Reply:
Please share when possible. If there's appetite for a larger refactor then perhaps that's the path that we should be taking. Is it something to raise in the mai-tai call?

Reply:
It was in collecting initial feedback stage. I will revamp it and share with you.